File: src/main/java/com/fengsheng/phase/AfterDieGiveCard.java
package com.fengsheng.phase;

import com.fengsheng.Fsm;
import com.fengsheng.Player;
import com.fengsheng.ResolveResult;
import com.fengsheng.card.Card;

/**
 * 死亡给三张牌后，判断是否有下一个人需要结算死亡给三张牌
 */
public record AfterDieGiveCard(WaitForDieGiveCard dieGiveCard) implements Fsm {
    @Override
    public ResolveResult resolve() {
        Player player = dieGiveCard.diedQueue.get(dieGiveCard.diedIndex);
        Card[] cards = player.getCards().values().toArray(new Card[0]);
        player.getGame().playerDiscardCard(player, cards);
        player.getGame().getDeck().discard(player.deleteAllMessageCards());
        for (Player p : player.getGame().getPlayers()) {
            p.notifyDie(player.location());
        }
        dieGiveCard.diedIndex++;
        return new ResolveResult(dieGiveCard, true);
    }
}

File: src/main/kotlin/phase/AfterDieGiveCard.kt
package com.fengsheng.phase

import com.fengsheng.DiscardCardEvent
import com.fengsheng.Fsm
import com.fengsheng.ResolveResult

/**
 * 死亡给三张牌后，判断是否有下一个人需要结算死亡给三张牌
 */
data class AfterDieGiveCard(val dieGiveCard: WaitForDieGiveCard) : Fsm {
    override val whoseTurn
        get() = dieGiveCard.whoseTurn

    override fun resolve(): ResolveResult {
        val player = dieGiveCard.diedQueue[dieGiveCard.diedIndex]
        val cards = player.cards.toList()
        player.game!!.playerDiscardCard(player, cards)
        val messageCards = player.messageCards.toList()
        player.messageCards.clear()
        player.game!!.deck.discard(messageCards)
        for (p in player.game!!.players) {
            p!!.notifyDie(player.location)
        }
        dieGiveCard.diedIndex++
        if (player.cards.isNotEmpty())
            player.game!!.addEvent(DiscardCardEvent(dieGiveCard.whoseTurn, player))
        return ResolveResult(dieGiveCard, true)
    }
}
